cross-chain/solana/solana-tbtc-anchor/programs/solana-tbtc-anchor/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+        require!(ctx.accounts.tbtc.minters > 0, TbtcError::NoMinters);
+        ctx.accounts.tbtc.minters -= 1;
+        Ok(())
+    }
+
+    pub fn add_guardian(ctx: Context<AddGuardian>) -> Result<()> {
+        let guardian_info = &mut ctx.accounts.guardian_info;
+        guardian_info.guardian = ctx.accounts.guardian.key();
+        guardian_info.bump = *ctx.bumps.get("guardian_info").unwrap();
+
+        ctx.accounts.tbtc.guardians += 1;
+        Ok(())
+    }
+
+    pub fn remove_guardian(ctx: Context<RemoveGuardian>, _minter: Pubkey) -> Result<()> {
+        require!(ctx.accounts.tbtc.guardians > 0, TbtcError::NoGuardians);
+        ctx.accounts.tbtc.guardians -= 1;
+        Ok(())
+    }
+
+    pub fn setup_mint(_ctx: Context<SetupMint>) -> Result<()> {
+        msg!("setup mint complete\n\n\n");
+        Ok(())
+    }
+
+    pub fn mint(ctx: Context<Mint>, amount: u64) -> Result<()> {
+        msg!("start mint\n\n\n");
+        let seed = b"tbtc-mint";
+        let key_seed = ctx.accounts.tbtc.key();
+        let mint_bump = ctx.accounts.tbtc.token_bump;
+        let signer: &[&[&[u8]]] = &[&[seed, key_seed.as_ref(), &[mint_bump]]];
+
+        let cpi_ctx = CpiContext::new_with_signer(
+            ctx.accounts.token_program.to_account_info(),
+            SplMintTo {
+                mint: ctx.accounts.tbtc_mint.to_account_info(),
+                to: ctx.accounts.recipient_account.to_account_info(),
+                authority: ctx.accounts.tbtc_mint.to_account_info(),
+            },
+            signer,
+        );
+        mint_to(cpi_ctx, amount)?;
+
+        Ok(())
+    }
+
+    pub fn pause(ctx: Context<Pause>) -> Result<()> {
+        ctx.accounts.tbtc.paused = true;
+        Ok(())
+    }
+
+    pub fn unpause(ctx: Context<Unpause>) -> Result<()> {
+        ctx.accounts.tbtc.paused = false;
+        Ok(())
+    }
+}
+
+#[account]
+#[derive(Default)]
+pub struct Tbtc {
+    creator: Pubkey,
+    token_mint: Pubkey,
+    token_bump: u8,
+    minters: u8,
+    guardians: u8,
+    paused: bool,
+}
+
+#[account]
+#[derive(Default)]
+pub struct MinterInfo {
+    minter: Pubkey,
+    bump: u8,
+}
+
+#[account]
+#[derive(Default)]
+pub struct GuardianInfo {
+    guardian: Pubkey,
+    bump: u8,
+}
+
+#[derive(Accounts)]
+pub struct Initialize<'info> {
+    // Use PDA for the mint address
+    // so we can sign for it from the program
+    #[account(
+        init,
+        seeds = [b"tbtc-mint", tbtc.key().as_ref()],
+        bump,
+        payer = creator,
+        mint::decimals = 8,
+        mint::authority = tbtc_mint,
+    )]
+    pub tbtc_mint: Account<'info, SplMint>,
+
+    #[account(init, payer = creator, space = 8 + Tbtc::MAXIMUM_SIZE)]
+    pub tbtc: Account<'info, Tbtc>,
+
+    #[account(mut)]
+    pub creator: Signer<'info>,
+    
+    pub token_program: Program<'info, SplToken>,
+    pub system_program: Program<'info, System>
+}
+
+#[derive(Accounts)]
+pub struct AddMinter<'info> {
```

Comment:
I'm a little confused about adding a minter functionality. On EVM implementation we have an array of minters that we can check against. Can you please shed some light on how this array is implemented here? Should we have an array of minters or is it just adding a new authority minter account for a mint and no array is needed to keep track of it?

Reply:
There is no array in this version; it only keeps track of the number of minters. The minters are stored in accounts derived from the minter's pubkey, but there is no provided way to iterate over them.

If we want to add iteration, we could construct it the following way:

```
pub struct TbtcInfo {
  minter_count: u32
}

// each minter index is derived only from the address of TBTC and the index,
// minter_index[x] = pda(tbtc_address, "minter-index", x)
// and contains the pubkey of the minter authority at that index
// (we don't need to store the minter account address as it's derived from the pubkey)
pub struct MinterIndex {
  minter_authority: Pubkey
}

// the minter account holds its index and the pubkey of the authority
pub struct MinterAccount {
  minter_index: u32
  authority_address: Pubkey
}
```

with this structure, you can iterate through all minters by querying the minter indices up to `minter_count - 1`; you don't need to know the address of each minter authority in advance as the addresses are derived only from the count

to add a minter, we increment `minter_count`, add a new `MinterAccount` and a new `MinterIndex` at the previous minter count (e.g. to add a minter when we have 3 minters, we set minter count to 4, and add a minter index at `pda(tbtc, "minter-index", 3)`)

to remove a minter, we decrement `minter_count`, remove the `MinterAccount`, and replace address in the removed minter's `MinterIndex` with the last `MinterIndex` address, then delete the last `MinterIndex`

and similar for guardians

Reply:
Ok, thank you for the explanation. I don't think we need to iterate through minters or guardians. We are only interested in functionality that an authorized minter(s) or guardian(s) can execute certain functions. We can leave it as is.